### ISSUE TITLE
member-ordering: Allow category grouping

### ISSUE THREAD

**Issue by DanielSWolf** [palantir/tslint#1912]
The `member-ordering` rule gives me a set of member categories (`public-static-field`, `protected-instance-method`...) and allows me to arrange them in any order. The problem with this approach is that I have to specify a *total order* -- that is, I can't have "gray areas" where I can mix members belonging to multiple categories.

There is at least one case where this is problematic: fields depending on one another. Here's an example:

```js
class Foo {
  private static implementationDetail = 42;
  public static bar = Foo.implementationDetail * 2;
}
```

In this case, I *have* to specify the private field `implementationDetail` before the public field `bar`. Otherwise, the resulting JavaScript code will try to initialize `bar` before `implementationDetail`, meaning that `bar` will be set to `undefined * 2`.

In practice, I may have public fields depending on private fields (as in this example), private fields depending on public fields, or even both. So there isn't a *total order* that can work.

What I'd love to see is a way to *group* categories, allowing me to mix members from all categories within a group. The result could look like this:

```js
"member-ordering": [true, { "order": [
    ["public-static-field", "protected-static-field", "private-static-field"],
    "public-static-method",
    "protected-static-method",
    "private-static-method",
    ["public-instance-field", "protected-instance-field", "private-instance-field"],
    "constructor",
    "public-instance-method",
    "protected-instance-method",
    "private-instance-method",
]}]
```

**Comment by aciccarello** [comment]
I think this was resolved

**Comment by billfranklin** [comment]
as of what version?

**Comment by aciccarello** [comment]
@billfranklin Sorry, I should have added more information.

The [rule documentation](https://palantir.github.io/tslint/rules/member-ordering/) demonstrates how to create your own categories. These should allow mixing order within the category.

> You can also make your own categories by using an object instead of a string:
> ```
> {
>     "name": "static non-private",
>     "kinds": [
>         "public-static-field",
>         "protected-static-field",
>         "public-static-method",
>         "protected-static-method"
>     ]
> }
> ```

**Comment by billfranklin** [comment]
@aciccarello I don't think that does what you think it does.  I believe what we are all wanting is the same functionality that ordered-imports has....with the caveat that groups of kinds would be used instead of groups of imports....but since a blank line exists already for most, perhaps a blank line and comment is required?
_from ordered-imports:_
"Groups of imports are delineated by blank lines. You can use these to group imports however you like, e.g. by first- vs. third-party or thematically."

_what we want:_
"Groups of kinds are delineated by a blank line and comment.  You can use these to group kinds however you like, e.g. thematically, etc."

**Comment by rkrisztian** [comment]
@billfranklin, comments weren't really part of the original request though. So to respond to @DanielSWolf, the documentation says:

> You can also omit the access modifier to refer to “public-“, “protected-“, and “private-“ all at once; for example, “static-field”.

**Comment by billfranklin** [comment]
I don't think that covers it though.  

We would want, for example, something like this to group related code for readability

```
// form data functions
getEquipment(locationId: number) {
    // code here
}
getLocations() {
    // code here
}

// form functions
resetForm() {
    // code here
}

submitForm() {
    // code here
}

// user functions
getUser(userId: number): User {
    // code here
}

updateUser(user: User) {
    // code here
}

```

**Comment by rkrisztian** [comment]
I see. Well, those comments look like banners that, according to the Clean Code book from Robert C. Martin, don't work in practice, because they're noisy and usually ignored. Also, as you pointed out, it can makes sense to group methods by functionalities (i.e. group related methods close together, which is described in the Clean Code book), in which case a rigid ordering on visibility modifiers can just get in the way, true. (That's why I don't define ordering on visibility either.) But I think this is too difficult to enforce automatically as it is as it is too difficult to compete with human intelligence. For one example, what defines the boundaries? Comments are free-text, and the language should naturally support method groups instead. Or what if groups are not so easy to define because they kind of overlap? Then there are multiple correct ways, thus maybe it's not worth to spend time on arguing about them. That's why I've never seen a linter tool that has a rule for grouping. Also, maybe it's that you're violating SRP too much, and you should separate your class content.  => So there are so many things to consider in such checks, so analysis results would be likely to point you to the wrong conclusion.
